lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -75,33 +98,92 @@ public String toString() {
+            this.executor = executor;
+        }
 
+        /**
+         * Schedule a task for concurrent execution.
+         * If there is insufficient room to schedule the task, safely, on the underlying
+         * executor, the task is added to the pendingTask queue and executed when spaces is available.
+         * There are absolutely no guarantees about execution order, on this executor,
+         * particularly once it fails back to using the pending task queue.
+         *
+         * @param task a task for concurrent execution.
+         * @throws ExecutorClosedException    if the executor has been stopped
+         * @throws RejectedExecutionException if the underlying executor rejects the task
+         */
         @Override
-        public synchronized void execute(@NonNull Runnable task) {
-            if (stopLatch != null) { throw new ExecutorClosedExeception("Executor has been stopped"); }
+        public void execute(@NonNull Runnable task) {
+            Preconditions.checkArgNotNull(task, "task");
 
-            try {
-                executor.execute(new InstrumentedTask(task, this::finishTask));
-                running++;
-            }
-            catch (RejectedExecutionException e) {
-                dumpServiceState(executor, e, "size: " + running);
-                throw e;
+            final int n;
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                final InstrumentedTask newTask;
+                if (spaceAvailable()) { newTask = new InstrumentedTask(task, this::finishTask); }
+                else {
+                    newTask = new InstrumentedTask(task, this::scheduleNext);
+                    pendingTasks.add(newTask);
+                }
+
+                n = pendingTasks.size();
+                if (n <= 1) { executeTask(newTask); }
             }
+
+            if (n > 0) { Log.w(DOMAIN, "Parallel executor overflow: " + n); }
         }
 
+        /**
+         * Stop the executor.
+         * If there are pending (unscheduled) tasks, they are abandoned.
+         * If this call returns false, the executor has *not* yet stopped: tasks it scheduled are still running.
+         *
+         * @param timeout time to wait for shutdown
+         * @param unit    time unit for shutdown wait
+         * @return true if all currently scheduled tasks have completed
+         */
         @Override
         public boolean stop(long timeout, @NonNull TimeUnit unit) {
+            Preconditions.testArg(timeout, "timeout must be >= 0", x -> x >= 0);
+            Preconditions.checkArgNotNull(unit, "time unit");
+
             final CountDownLatch latch;
             synchronized (this) {
-                if (stopLatch == null) { stopLatch = new CountDownLatch(running); }
+                if (stopLatch == null) {
+                    pendingTasks.clear();
+                    stopLatch = new CountDownLatch(1);
```

Comment:
Why was this changed to 1?  It seems like this would get over counted down.

Reply:
Because of the change in line 200 that will only count down the latch when all running tasks are done?